crates/starknet-types-core/src/felt/qm31.rs

```diff
@@ -0,0 +1,489 @@
+use core::fmt;
+
+use crate::felt::Felt;
+
+pub const STWO_PRIME: u64 = (1 << 31) - 1;
+const STWO_PRIME_U128: u128 = STWO_PRIME as u128;
+const MASK_36: u64 = (1 << 36) - 1;
+const MASK_8: u64 = (1 << 8) - 1;
+
+#[derive(Debug)]
+pub enum QM31Error {
+    UnreducedFelt(Felt),
+    FeltTooBig(Felt),
+    InvalidInversion,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for QM31Error {}
+
+impl fmt::Display for QM31Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            QM31Error::UnreducedFelt(felt) => write!(
+                f,
+                "Number is not a packing of a QM31 in reduced form: {felt})"
+            ),
+            QM31Error::FeltTooBig(felt) => write!(
+                f,
+                "Number used as QM31 since it's more than 144 bits long: {felt}"
+            ),
+            QM31Error::InvalidInversion => write!(f, "Attempt to invert a qm31 equal to zero"),
+        }
+    }
+}
+
+/// Definition of a Quad M31 in its reduced form. The internal representation
+/// is composed by the coordinates of the QM31, following a little endian ordering.
+#[repr(transparent)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct QM31([u64; 4]);
+
+impl QM31 {
+    /// [QM31] constant that's equal to 0.
+    pub const ZERO: Self = Self([0, 0, 0, 0]);
+
+    pub fn inner(&self) -> [u64; 4] {
+        self.0
+    }
+
+    /// Creates a [QM31] in its reduced form from four u64 coordinates.
```

Comment:
Can you provide a link to what `coordinates` means in this context? I'm guessing it is some well known mathematical concept.